### PR TITLE
[Model Monitoring] Change controller from scheduled job to daemon

### PR DESCRIPTION
**Pull request by davesh0812**
[ML-5149](https://jira.iguazeng.com/browse/ML-5149)

Few extra changes : 
1. Daemon controller deployed with 2 kind of triggers : Http trigger & CronTrigger with interval according to the base_period - [ML-5828](https://jira.iguazeng.com/browse/ML-5828)
3. Update Api for enable model monitoring - [ML-5825](https://jira.iguazeng.com/browse/ML-5825)
4. Update Api for disable model monitoring - delete only from db (not from nuclio - blocked by https://jira.iguazeng.com/browse/ML-3432).
5. New Api for update model monitoring controller -[ML-5827](https://jira.iguazeng.com/browse/ML-5827)
6. Controller, writer and stream functions deployed when running `project.enable_model_monitoring()`
7. Only previous batch job created when running deploy on first serving function (with set_tracking).

### CODE REVIEW

File: server/api/api/endpoints/jobs.py [hunk 1]
```diff
@@ -106,86 +107,3 @@ async def deploy_monitoring_batch_job(
...
-        project_name=project,
-        resource_name=mlrun.common.schemas.model_monitoring.MonitoringFunctionNames.APPLICATION_CONTROLLER,
-        action=mlrun.common.schemas.AuthorizationAction.store,
-        auth_info=auth_info,
-    )
-
-    if (
-        mlrun.mlconf.httpdb.clusterization.role
-        != mlrun.common.schemas.ClusterizationRole.chief
-    ):
```
Comment (Eyal-Danieli) [inline server/api/api/endpoints/jobs.py:150]: Since we are no longer submit a scheduled job, we don't need to reroute this API the router. In other words, please remove the `create_model_monitoring_controller` function from `server/api/utils/clients/chief.py` https://github.com/mlrun/mlrun/blob/e0b6e66ccd3524ec130451c19c46be5557ded06b/server/api/utils/clients/chief.py#L215

File: mlrun/model_monitoring/controller.py [hunk 2]
```diff
@@ -282,33 +282,35 @@ class MonitoringApplicationController:
...
+                    mm_constants.EventFieldType.BATCH_INTERVALS_DICT
+                )
+            )
         )
 
         # If provided, only model endpoints in that that list will be analyzed
-        self.model_endpoints = context.parameters.get(
-            mm_constants.EventFieldType.MODEL_ENDPOINTS, None
-        )
+        self.model_endpoints = None
```
Comment (Eyal-Danieli) [inline mlrun/model_monitoring/controller.py:313]: Not sure why do we need that `self._model_endpoints = None` here. How should the user provide this list? When the controller was a scheduled job I can understand that perhaps we would like to provide a list of model endpoints to process. However, now it's a real-time function. As I see it we have 4 options:
1 - use a flag/label on the tracked model endpoints
2 - provide optional model endpoints list through the incoming event to the controller handler
3 - store model endpoints list as an env/secret (seems unnecessary) 
4 - manage the processed model endpoints through the monitoring policy (once implemented)

Anyway, at the moment I think we can remove it from the controller code and implement it as a different PR with the selected approach.

File: server/api/api/endpoints/model_monitoring.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+from server.api.api.endpoints.functions import process_model_monitoring_secret
+from server.api.crud.model_monitoring.deployment import MonitoringDeployment
+
+router = fastapi.APIRouter(prefix="/projects/{project}/model-monitoring")
+
+
+@router.post("/enable-model-monitoring")
+async def enable_model_monitoring(
+    project: str,
+    request: fastapi.Request,
```
Comment (Eyal-Danieli) [inline server/api/api/endpoints/model_monitoring.py:32]:
```suggestion
```
We used it for routing the request to the chief instance which is not longer required

File: server/api/crud/model_monitoring/deployment.py [hunk 4]
```diff
@@ -398,7 +467,7 @@ def _initial_model_monitoring_stream_processing_function(
         self,
         project: str,
         model_monitoring_access_key: str,
-        tracking_policy: mlrun.model_monitoring.tracking_policy.TrackingPolicy,
```
Comment (Eyal-Danieli) [inline server/api/crud/model_monitoring/deployment.py:401]: remove from docstring as well

**Review by Eyal-Danieli — COMMENTED** [review]
nice
minor comments + need to add tests (validating the controller deployment, validating the base period...)

File: server/api/crud/model_monitoring/deployment.py [hunk 5]
```diff
@@ -177,41 +196,90 @@ def deploy_model_monitoring_controller(
...
+        }
+        fn.set_env(
+            mm_constants.EventFieldType.BATCH_INTERVALS_DICT,
+            json.dumps(batch_dict),
+        )
+
+        fn.add_trigger(
+            "cron_interval",
+            spec=nuclio.CronTrigger(interval=f"{base_period}m"),
+        )
```
Comment (jond01) [inline server/api/crud/model_monitoring/deployment.py:277]: Looks neat!
To make sure I understand, trigger events will be sent according to the set interval, but the function is always up - a daemon. Right?

**Review by jond01 — COMMENTED** [review]
Looks good!
I had some questions and comments

File: server/api/api/endpoints/jobs.py [hunk 6]
```diff
@@ -106,86 +107,3 @@ async def deploy_monitoring_batch_job(
...
-        logger.info(
-            "Requesting to deploy model monitoring controller, re-routing to chief",
-            function_name="model-monitoring-controller",
-            project=project,
-        )
-        chief_client = server.api.utils.clients.chief.Client()
```
[the remark this answers is not in the document]
Comment (jond01) [reply server/api/api/endpoints/jobs.py:156]: Eyal's comment https://github.com/mlrun/mlrun/pull/5178#discussion_r1497286373 probably answers this question

File: mlrun/model_monitoring/controller_handler.py [hunk 7]
```diff
@@ -11,19 +11,27 @@
...
+    mlrun_context = mlrun.get_or_create_ctx("model_monitoring_controller")
+    mlrun_context.logger.info("Initialize monitoring app controller")
     monitor_app_controller = MonitoringApplicationController(
-        context=context,
-        project=context.project,
+        mlrun_context=mlrun_context,
+        project=mlrun_context.project,
     )
-    monitor_app_controller.run()
+    setattr(context.user_data, "monitor_app_controller", monitor_app_controller)
```
[the remark this answers is not in the document]
Comment (davesh0812) [reply mlrun/model_monitoring/controller_handler.py:37]: It is the best practice of nuclio,  You have another suggestion ?

File: mlrun/model_monitoring/controller_handler.py [hunk 8]
```diff
@@ -11,19 +11,27 @@
...
 from mlrun.model_monitoring.controller import MonitoringApplicationController
 
 
-def handler(context: mlrun.run.MLClientCtx) -> None:
+def handler(context: nuclio.Context, event: nuclio.Event) -> None:
     """
     Run model monitoring application processor
 
     :param context: the MLRun context
+    :param event:   trigger event
```
[the remark this answers is not in the document]
Comment (davesh0812) [reply mlrun/model_monitoring/controller_handler.py:25]: Not now, but I will add it because we will need it for the offline batch infer

File: server/api/api/endpoints/functions.py [hunk 9]
```diff
@@ -832,15 +833,17 @@ def _deploy_serving_monitoring(
                 ),
             )
 
+        # TODO : delete when batch is deprecated.
         # deploy model monitoring stream, model monitoring batch job,
```
[the remark this answers is not in the document]
Comment (davesh0812) [reply server/api/api/endpoints/functions.py:837]: https://github.com/mlrun/mlrun/pull/5178#discussion_r1497183412 will be update instead

File: mlrun/db/httpdb.py [hunk 10]
```diff
@@ -3076,31 +3076,60 @@ def create_model_monitoring_controller(
         project: str = "",
         default_controller_image: str = "mlrun/mlrun",
         base_period: int = 10,
+        overwrite: bool = False,
```
[the remark this answers is not in the document]
Comment (davesh0812) [reply mlrun/db/httpdb.py:3079]: https://github.com/mlrun/mlrun/pull/5178#discussion_r1497183412  - it will be `update_model_monitoring_controller ` instead.

File: server/api/crud/model_monitoring/deployment.py (hunk 5 above)
Comment (davesh0812) [reply server/api/crud/model_monitoring/deployment.py:277]: Yep :)

File: mlrun/model_monitoring/controller_handler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jond01) [reply mlrun/model_monitoring/controller_handler.py:37]: Let's keep it as it then

File: server/api/crud/model_monitoring/deployment.py [hunk 11]
```diff
@@ -862,6 +805,38 @@ def _initial_model_monitoring_writer_function(
...
+        """
+        if not overwrite:
+            logger.info(
+                f"Checking if {function_name} is already deployed",
+                project=project,
+            )
+            try:
+                # validate that the model monitoring stream has not yet been deployed
+                mlrun.runtimes.nuclio.function.get_nuclio_deploy_status(
+                    name=mm_constants.MonitoringFunctionNames.WRITER,
```
Comment (jond01) [inline server/api/crud/model_monitoring/deployment.py:826]: Shouldn't it be `function_name`?

File: server/api/crud/model_monitoring/deployment.py [hunk 12]
```diff
@@ -218,30 +197,11 @@ def deploy_model_monitoring_controller(
...
-                    "Detected model monitoring controller processing function already deployed",
-                    project=project,
-                )
-                return
-            except mlrun.errors.MLRunNotFoundError:
-                pass
-        logger.info(
-            "Deploying model monitoring controller processing function",
+        MonitoringDeployment._check_if_already_deployed(
+            function_name=mm_constants.MonitoringFunctionNames.APPLICATION_CONTROLLER,
```
Comment (jond01) [inline server/api/crud/model_monitoring/deployment.py:201]: I think it doesn't check the controller - see the previous comment.

File: mlrun/projects/project.py [hunk 13]
```diff
@@ -2009,6 +2007,31 @@ def enable_model_monitoring(
...
             base_period=base_period,
+        )
+
+    def update_model_monitoring_controller(
+        self,
+        base_period: int = 10,
+        overwrite: bool = False,
+        image: str = "mlrun/mlrun",
+    ) -> dict:
+        r"""
```
Comment (jond01) [inline mlrun/projects/project.py:2018]: Why `r"""`?

File: server/api/api/endpoints/model_monitoring.py [hunk 14]
```diff
@@ -131,13 +122,27 @@ async def deploy_model_monitoring_controller(
...
+        mlrun.runtimes.nuclio.function.get_nuclio_deploy_status(
+            name=mlrun.common.schemas.model_monitoring.MonitoringFunctionNames.APPLICATION_CONTROLLER,
+            project=project,
+            tag="",
+            auth_info=auth_info,
+        )
+
+    except mlrun.errors.MLRunNotFoundError:
+        raise mlrun.errors.MLRunNotFoundError(
+            f"{mlrun.common.schemas.model_monitoring.MonitoringFunctionNames.APPLICATION_CONTROLLER} is not exists."
```
Comment (jond01) [inline server/api/api/endpoints/model_monitoring.py:136]: Change "is not exists." -> "does not exist" :)

File: mlrun/db/httpdb.py [hunk 15]
```diff
@@ -3074,36 +3074,59 @@ def deploy_monitoring_batch_job(
...
-                                         image. By default, the image is mlrun/mlrun.
-        :param base_period:              Minutes to determine the frequency in which the model monitoring controller job
-                                         is running. By default, the base period is 5 minutes.
-        :returns: model monitoring controller job as a dictionary. You can easily convert the returned function into a
-                  runtime object by calling ~mlrun.new_function.
+        :param image: The image of the model monitoring controller function.
+                                         By default, the image is mlrun/mlrun.
+        :param base_period:              The time period in minutes in which the model monitoring controller function
+                                         triggers. By default, the base period is 10 minutes.
+        :param overwrite:                If true, overwrite the existing model monitoring controller. By default, False.
```
Comment (Eyal-Danieli) [inline mlrun/db/httpdb.py]:
```suggestion

```

remove `overwrite` from docstring

File: server/api/api/endpoints/functions.py [hunk 16]
```diff
@@ -809,6 +818,7 @@ def _deploy_serving_monitoring(
     model_monitoring_access_key,
     monitoring_application,
```
Comment (Eyal-Danieli) [inline server/api/api/endpoints/functions.py]: remove unused parameter `monitoring_application` from function signature

File: server/api/api/endpoints/functions.py [hunk 17]
```diff
@@ -789,8 +791,15 @@ def _deploy_nuclio_runtime(
                 monitoring_application,
```
Comment (Eyal-Danieli) [inline server/api/api/endpoints/functions.py]:
```suggestion
```
unused parameter

**Review by Eyal-Danieli — APPROVED** [review]
LGTM
just need to remove unused parameters

File: mlrun/common/schemas/model_monitoring/constants.py [hunk 18]
```diff
@@ -181,7 +181,7 @@ class MonitoringFunctionNames:
     WRITER = "model-monitoring-writer"
     BATCH = "model-monitoring-batch"
     APPLICATION_CONTROLLER = "model-monitoring-controller"
-    STREAM = None
+    STREAM = "model-monitoring-stream"
```
Comment (jond01) [inline mlrun/common/schemas/model_monitoring/constants.py:184]: Nice! I tried to do that in the past but it didn't work for me

File: mlrun/common/model_monitoring/helpers.py [hunk 19]
```diff
@@ -72,7 +73,10 @@ def parse_monitoring_stream_path(
                 "Custom kafka topic is not allowed"
             )
         # Add topic to stream kafka uri
-        if application_name is None:
+        if (
+            application_name is None
+            or application_name == mm_constants.MonitoringFunctionNames.STREAM
+        ):
```
Comment (jond01) [inline mlrun/common/model_monitoring/helpers.py]: In that case, I think renaming `application_name` to `function_name`, as the monitoring stream is not an application in out terminology

File: mlrun/model_monitoring/helpers.py [hunk 20]
```diff
@@ -41,19 +42,22 @@ class _MLRunNoRunsFoundError(Exception):
     pass
 
 
-def get_stream_path(project: str = None, application_name: str = None):
+def get_stream_path(
+    project: str = None,
+    application_name: str = mm_constants.MonitoringFunctionNames.STREAM,
```
Comment (jond01) [inline mlrun/model_monitoring/helpers.py]: See that previous comment about the naming of this argument.
Also, what happens if users try to create a model monitoring function named `mm_constants.MonitoringFunctionNames.STREAM`?

File: mlrun/projects/project.py (hunk 13 above)
Comment (jond01) [reply mlrun/projects/project.py]: Not sure that I understand the response here. I think that the r-string is unnecessary

File: tests/system/model_monitoring/test_app.py [hunk 21]
```diff
@@ -349,7 +350,8 @@ def _record_results(self) -> None:
             model_endpoint_name=f"{self.name_prefix}-test",
             function_name=self.function_name,
             endpoint_id=self.endpoint_id,
-            context=mlrun.get_or_create_ctx(name=f"{self.name_prefix}-context"),  # pyright: ignore[reportGeneralTypeIssues]
+            context=mlrun.get_or_create_ctx(name=f"{self.name_prefix}-context"),
+            # pyright: ignore[reportGeneralTypeIssues]
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]: Why? This should be on the same line - please revert this change

File: tests/system/model_monitoring/test_app.py [hunk 22]
```diff
@@ -371,3 +373,54 @@ def test_inference_feature_set(self) -> None:
         time.sleep(2.4 * self.app_interval_seconds)
 
         self._test_v3io_records(self.endpoint_id)
+
+
+@TestMLRunSystem.skip_test_if_env_not_configured
+@pytest.mark.enterprise
+class TestModelMonitoringInitialize(TestMLRunSystem, _V3IORecordsChecker):
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]: You don't need the second inheritance:
```suggestion
class TestModelMonitoringInitialize(TestMLRunSystem):
```

File: server/api/api/endpoints/model_monitoring.py [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+        mlrun.runtimes.nuclio.function.get_nuclio_deploy_status(
+            name=mlrun.common.schemas.model_monitoring.MonitoringFunctionNames.APPLICATION_CONTROLLER,
+            project=project,
+            tag="",
+            auth_info=auth_info,
+        )
+
+    except mlrun.errors.MLRunNotFoundError:
+        raise mlrun.errors.MLRunNotFoundError(
+            f"{mlrun.common.schemas.model_monitoring.MonitoringFunctionNames.APPLICATION_CONTROLLER} does not exists. "
```
Comment (jond01) [inline server/api/api/endpoints/model_monitoring.py]:
```suggestion
            f"{mlrun.common.schemas.model_monitoring.MonitoringFunctionNames.APPLICATION_CONTROLLER} does not exist. "
```
https://github.com/mlrun/mlrun/pull/5178#discussion_r1499176614
🙂

File: mlrun/model_monitoring/helpers.py (hunk 20 above)
Comment (davesh0812) [reply mlrun/model_monitoring/helpers.py]: @jond01 I think we have also a problem with the names of the controller and the writer.
I can restricted using those names in `set_model_monitoring_function`, make sense?

Comment (jond01) [reply mlrun/model_monitoring/helpers.py]: Yes, it sounds ok. Thanks

**Review by jond01 — APPROVED** [review]
Please just address the open comments, apart from that it looks good

File: tests/system/model_monitoring/test_app.py [hunk 24]
```diff
@@ -349,7 +350,6 @@ def _record_results(self) -> None:
             model_endpoint_name=f"{self.name_prefix}-test",
             function_name=self.function_name,
             endpoint_id=self.endpoint_id,
-            context=mlrun.get_or_create_ctx(name=f"{self.name_prefix}-context"),  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]: Keep that 😰

Comment (davesh0812) [reply tests/system/model_monitoring/test_app.py]: Ah oops
